### DISCUSSION TITLE
optional filed with self-relation error

### DISCUSSION THREAD

**Discussion by leohxj** [prisma/prisma#2542] [Q&A] [answered]
## Bug description
I want to define a parent field which means it is a re-run analyze item, so it relate to itself and optional.

here is schema.graphql define and errors:
![image](https://user-images.githubusercontent.com/1506900/82789539-0d650600-9e9d-11ea-9394-3734ff1490db.png)

another question when I want change parent as a required filed, after formatted, it auto generate an User filed:
![image](https://user-images.githubusercontent.com/1506900/82789740-6896f880-9e9d-11ea-9491-4cefe2a99799.png)

## How to reproduce
two case shcema is here:
```
datasource db {
  provider = "postgresql"
  url      = env("POSTGRESQL_URL")
}

generator client {
  provider = "prisma-client-js"
}

model Analyze {
  id       Int      @id @default(autoincrement())
  name     String?
  parent  Analyze?      @relation("AnalyzePartners")
}

```
and

```
datasource db {
  provider = "postgresql"
  url      = env("POSTGRESQL_URL")
}

generator client {
  provider = "prisma-client-js"
}

model Analyze {
  id       Int      @id @default(autoincrement())
  name     String?
  parent  Analyze      @relation("AnalyzePartners", fields: [id], references: [id])
}

```

## Prisma information
version of `@prisma/client` and `@prisma/cli` is    "2.0.0-beta.5"

## Environment & setup

- OS: [e.g. Mac OS, Windows, Debian, CentOS, ...]
Mac OS
- Database: [PostgreSQL, MySQL, MariaDB or SQLite]
PostgreSQL
- Prisma version: [Run `prisma -v` to see your Prisma version]
2.0.0-beta.5
- Node.js version: [Run `node -v` to see your Node.js version]
v10.15.0

**Comment by pantharshit00** [comment, marked as answer]
For the first case, you will need to provide the relation a field so that we can store the actual foreign key on it. All relations in the schema are virtual, you will need to create a field which backs the relation:

```prisma
model Analyze {
  id       Int       @default(autoincrement()) @id
  name     String?
  parent   Analyze?  @relation("AnalyzePartners", fields: [parentId], references: [id])
  parentId Int?
  childs   Analyze[] @relation("AnalyzePartners")
}
```

Also, please note the a relation need a back side in order determine the type of relation. Since the backing relation here has an list, this will be a 1:m relation.

Read: https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-schema/relations#self-relations

Now for the second question, auto formatting also automatically creates a back relation for you if it is not there. As explained you need a back relation so that we can determine its type and you can also query that. 


I will recommend you go through this docs page: https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-schema/relations#self-relations

**Reply by leohxj** [reply]
about 1:1 self-relation, is it must define two @relation? such as:

```
model User {
  id             Int     @default(autoincrement()) @id
  name           String?
  successorId Int?
  successor   User?   @relation("BlogOwnerHistory", fields: [successorId], references: [id])
  predecessor User?   @relation("BlogOwnerHistory")
}
```

I expected it as：
```
model User {
  id             Int     @default(autoincrement()) @id
  name           String?
  successorId Int?
  successor   User?   @relation("BlogOwnerHistory", fields: [successorId], references: [id])
}
```

does predecesor field necessary?

**Reply by nikolasburk** [reply]
Hey @leohxj, yes a relation always requires both fields to be present in the Prisma schema!

**Comment by ludralph** [comment]
Hi @leohxj,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
